src/Features/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/DotnetCliHelper.cs

```diff
@@ -0,0 +1,172 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Composition;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.InteropServices;
+using System.Text;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.Extensions.Logging;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.LanguageServer;
+
+[Export, Shared]
+internal class DotnetCliHelper
+{
+    internal const string DotnetRootEnvVar = "DOTNET_ROOT";
+    internal const string DotnetRootUserEnvVar = "DOTNET_ROOT_USER";
+
+    private readonly ILogger _logger;
+    private readonly Lazy<string> _dotnetExecutablePath;
+    private readonly AsyncLazy<string> _dotnetSdkFolder;
+
+    [ImportingConstructor]
+    [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+    public DotnetCliHelper(ILoggerFactory loggerFactory)
+    {
+        _logger = loggerFactory.CreateLogger<DotnetCliHelper>();
+        _dotnetExecutablePath = new Lazy<string>(() => GetDotnetExecutablePath());
+        _dotnetSdkFolder = new AsyncLazy<string>(GetDotnetSdkFolderFromDotnetExecutableAsync, cacheResult: true);
+    }
+
+    public string GetDotnetExecutablePath()
+    {
+        // The client can modify the value in DOTNET_ROOT to ensure that the server starts on the correct runtime.
+        // It will save the user's DOTNET_ROOT in the DOTNET_ROOT_USER environment variable, so check there instead of using the DOTNET_ROOT.
+        var dotnetPathFromDotnetRootUser = Environment.GetEnvironmentVariable(DotnetRootUserEnvVar);
+        if (TryGetDotnetExecutableFromFolder(dotnetPathFromDotnetRootUser, DotnetRootUserEnvVar, _logger, out var dotnetRootUserExecutable))
+        {
+            return dotnetRootUserExecutable;
```

Comment:
log here?  up to you.

Reply:
This logs inside TryGetDotnetExecutableFromFolder